source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2058 @@
+        }
+        aws_mutex_unlock(&nw_socket->synced_data.lock);
+    }
+
+    s_socket_release_internal_ref(nw_socket);
+    aws_mem_release(connection_args->allocator, task);
+    aws_mem_release(connection_args->allocator, connection_args);
+}
+
+static void s_schedule_connection_state_changed_fn(
+    struct aws_socket *socket,
+    struct nw_socket *nw_socket,
+    nw_connection_t nw_connection,
+    nw_connection_state_t state,
+    nw_error_t error) {
+
+    AWS_LOGF_TRACE(AWS_LS_IO_SOCKET, "id=%p: s_schedule_connection_state_changed_fn start...", (void *)nw_socket);
+
+    aws_mutex_lock(&nw_socket->synced_data.lock);
+
+    if (s_validate_event_loop(nw_socket->event_loop)) {
+        struct aws_task *task = aws_mem_calloc(nw_socket->allocator, 1, sizeof(struct aws_task));
+
+        struct connection_state_change_args *args =
+            aws_mem_calloc(nw_socket->allocator, 1, sizeof(struct connection_state_change_args));
+
+        args->socket = socket;
+        args->nw_socket = nw_socket;
+        args->allocator = nw_socket->allocator;
+        args->error = error ? nw_error_get_error_code(error) : 0;
+        args->state = state;
+        args->nw_connection = nw_connection;
+
+        s_socket_acquire_internal_ref(nw_socket);
+
+        aws_task_init(task, s_process_connection_state_changed_task, args, "ConnectionStateChangedTask");
+
+        aws_event_loop_schedule_task_now(nw_socket->event_loop, task);
+        aws_mutex_unlock(&nw_socket->synced_data.lock);
+
+    } else if (state == nw_connection_state_cancelled) {
+        // If event loop is destroyed, no io events will be proceeded. Closed the internal socket.
+        aws_mutex_unlock(&nw_socket->synced_data.lock);
+        s_socket_release_internal_ref(nw_socket);
+
+    } else {
+        aws_mutex_unlock(&nw_socket->synced_data.lock);
+    }
+}
+
+static void s_process_listener_success_task(struct aws_task *task, void *args, enum aws_task_status status) {
+    (void)status;
+    struct nw_listener_connection_args *task_args = args;
+    struct aws_allocator *allocator = task_args->allocator;
+    struct nw_socket *listener_nw_socket = task_args->nw_socket;
+
+    if (status == AWS_TASK_STATUS_RUN_READY) {
+
+        if (listener_nw_socket) {
+            aws_mutex_lock(&listener_nw_socket->synced_data.lock);
+            struct aws_socket *listener = listener_nw_socket->synced_data.base_socket;
+
+            struct aws_socket *new_socket = aws_mem_calloc(allocator, 1, sizeof(struct aws_socket));
+            struct aws_socket_options options = listener->options;
+            int error = aws_socket_init(new_socket, allocator, &options);
+            if (error) {
+                aws_mem_release(allocator, new_socket);
+                nw_release(task_args->new_connection);
+                if (listener->accept_result_fn) {
+                    listener->accept_result_fn(listener, task_args->error_code, NULL, task_args->user_data);
```

Comment:
Why wouldn't we use `get_last_error()` as the error code here? Presumably task_args->error_code is 0 at this point?